src/Main.hs

```diff
@@ -80,27 +69,30 @@ trace (Scene objects lights camera config) x y =
     in
         case maybeIntersectedObject of
             Nothing -> backgroundColor
-            (Just intersectionObj@(_,object)) ->
+            Just intersectionObj  ->
                 getColorFromIntersection backgroundColor ray lights objects intersectionObj
 
-isLightVisible :: [Object] -> Vector -> Light -> Bool
-isLightVisible objects point light =
-    let
-        direction = normalize $ (center light) `sub` point
-        ray = Ray point direction
-        objIntersections = fmap (minIntersection ray) objects
-    in
-        all isNothing objIntersections
 
 getColorFromIntersection :: Color -> Ray -> [Light] -> [Object] -> (Double, Object) -> Color
-getColorFromIntersection defaultColor ray lights objects (hitDistance , hitObject) =
+getColorFromIntersection defaultColor ray lights objects (hitDistance, hitObject) =
     let
         hitPoint = pointAlongRay ray hitDistance
         otherObjects = filter (/= hitObject) objects
         visibleLights = filter (isLightVisible otherObjects hitPoint) lights
     in
         lambertColor hitPoint hitObject visibleLights
 
+isLightVisible :: [Object] -> Vector -> Light -> Bool
+isLightVisible objects point light =
+    let
+        toLightVector = center light `sub` point
+        distanceToLight = magnitude toLightVector
+        direction = normalize toLightVector
+        ray = Ray point direction
+        objIntersections = mapMaybe (minIntersection ray) objects
+    in
+        all ((\x -> x >= distanceToLight || x < 0) . fst) objIntersections
```

Comment:
@rumblesan these were the two cases we missed - the one where the light is behind the ray (ignored for the plane but not for the sphere, might look into that) and the one where the light is in front of the ray.